DSharpPlus/Clients/DiscordClient.Dispatch.cs

```diff
@@ -1589,6 +1677,69 @@ internal async Task OnMessageReactionRemoveEmojiAsync(ulong messageId, ulong cha
 
         #endregion
 
+        #region Stage Instance
+
+        internal async Task OnStageInstanceCreateEventAsync(DiscordStageInstance stage)
+        {
+            stage.Discord = this;
+
+            await this._stageInstanceCreated.InvokeAsync(this, new StageInstanceCreateEventArgs { StageInstance = stage }).ConfigureAwait(false);
+        }
+
+        internal async Task OnStageInstanceUpdateEventAsync(DiscordStageInstance stage)
+        {
+            stage.Discord = this;
+
+            await this._stageInstanceUpdated.InvokeAsync(this, new StageInstanceUpdateEventArgs { StageInstance = stage }).ConfigureAwait(false);
+        }
+
+        internal async Task OnStageInstanceDeleteEventAsync(DiscordStageInstance stage)
+        {
+            stage.Discord = this;
+
+            await this._stageInstanceDeleted.InvokeAsync(this, new StageInstanceDeleteEventArgs { StageInstance = stage }).ConfigureAwait(false);
+        }
+
+        #endregion
+
+        #region Thread
+
+        internal async Task OnThreadCreateEventAsync(DiscordThreadChannel thread)
+        {
+            thread.Discord = this;
+
+            await this._threadCreated.InvokeAsync(this, new ThreadCreateEventArgs { ThreadChannel = thread }).ConfigureAwait(false);
+        }
+
+        internal async Task OnThreadUpdateEventAsync(DiscordThreadChannel thread)
+        {
+            thread.Discord = this;
+
+            await this._threadUpdated.InvokeAsync(this, new ThreadUpdateEventArgs { ThreadChannel = thread }).ConfigureAwait(false);
+        }
+
+        internal async Task OnThreadDeleteEventAsync(DiscordThreadChannel thread)
+        {
+            thread.Discord = this;
+
+            await this._threadDeleted.InvokeAsync(this, new ThreadDeleteEventArgs { Id = thread.Id, GuildId = thread.GuildId, ParentChannelId = thread.ParentChannelId, Type = thread.Type }).ConfigureAwait(false);
```

Comment:
Use concrete DiscordGuild and DiscordChannel objects from the cache, don't pass an ID

Reply:
Uses concrete DiscordGuild and DiscordChannel for parent. Still passing on the thread id however as we don't currently keep a cache of DiscordThreadChannel and the payload is only the thread id, guild id, parent id, and that it was a thread.